### DISCUSSION TITLE
How to optimize dxf path

### DISCUSSION THREAD

**Discussion by lypwig** [mozman/ezdxf#750] [Q&A]
I have a dxf file not optimized, in a sense that there are many useless moves:

![image](https://user-images.githubusercontent.com/1665542/194731478-62bb88f6-8539-4cea-a89f-bcf2c7c2b088.png)

On the image above, only circles and the contour belongs to the actual dxf, other lines are just moves.

I wanted to avoid this, so I wrote a basic path optimizer for ezdxf:

![image](https://user-images.githubusercontent.com/1665542/194731533-622ce00d-f1f0-4cbc-9dee-b61d049de2d1.png)

It's not perfect, but before to improve it I wonder if there is a built-in ezdxf feature to do this.

If not, I can improve and publish my script as an ezdxf addon.

**Comment by mozman** [comment]
Thank you for your offer, but in my opinion this is a task for a human using an interactive CAD application and cannot be solved by an automated function call of a library like ezdxf. 

Even with a simple task like this, it's hard to tell what the "correct" result is:

![image](https://user-images.githubusercontent.com/4230966/194738780-86b62776-5d10-49c4-9446-eef3ae01c124.png)

**Reply by lypwig** [reply]
Hmm, Maybe I am misunderstood. I don't want to remove "cuts", I want to reduce "moves" (or, in other words, reorder dxf entities). So in your example above, the result will be the original file.

When I import a dxf file to the laser cutter I use, the tool doesn't optimize paths and takes more time to move than to actually cut. I'm not going to fix this by hand, because is a huge waste of time and it can be done automatically.

**Comment by mozman** [comment]
Shouldn't this be done by the cutter software?  In any case, this is a very specific problem for laser cutting and only marginally related to the DXF file format itself (the graphical entities look the same in a CAD application). 

BTW: CAD applications order the entities by ascending handle values by default. The entity order can be changed by the header variable `$SORTENTS` or adding a `SORTENTSTABLE` entity to the layout. Third party software often ignore this feature, most likely also the cutter software you are using.

- example for altering the redraw order: https://github.com/mozman/ezdxf/blob/master/examples/redraw_order.py
- docs for the reorder module: https://ezdxf.mozman.at/docs/reorder.html
